src/engine/x86-64/X86_64Interpreter.v3

```diff
@@ -1521,6 +1550,186 @@ class X86_64InterpreterGen(ic: X86_64InterpreterCode, w: DataWriter) {
+		bindHandler(Opcode.I32_ATOMIC_RMW8_SUB_U);
+		bindHandler(Opcode.I64_ATOMIC_RMW8_SUB_U);
+		genAtomicSub(asm.xaddb_m_r, asm.negb_r);
+		bindHandler(Opcode.I32_ATOMIC_RMW16_SUB_U);
+		bindHandler(Opcode.I64_ATOMIC_RMW16_SUB_U);
+		genAtomicSub(asm.xaddw_m_r, asm.negw_r);
+		// Atomic RMW And 
+		bindHandler(Opcode.I32_ATOMIC_RMW_AND);
+		bindHandler(Opcode.I64_ATOMIC_RMW32_AND_U);
+		genAtomicBinop(asm.andd_m_r, asm.xchgd_m_r);
+		bindHandler(Opcode.I64_ATOMIC_RMW_AND);
+		genAtomicBinop(asm.andq_m_r, asm.xchgq_m_r);
+		bindHandler(Opcode.I32_ATOMIC_RMW8_AND_U);
+		bindHandler(Opcode.I64_ATOMIC_RMW8_AND_U);
+		genAtomicBinop(asm.andb_m_r, asm.xchgb_m_r);
+		bindHandler(Opcode.I32_ATOMIC_RMW16_AND_U); 
+		bindHandler(Opcode.I64_ATOMIC_RMW16_AND_U);
+		genAtomicBinop(asm.andw_m_r, asm.xchgw_m_r);
+		// Atomic RMW Or
+		bindHandler(Opcode.I32_ATOMIC_RMW_OR);
+		bindHandler(Opcode.I64_ATOMIC_RMW32_OR_U);
+		genAtomicBinop(asm.ord_m_r, asm.xchgd_m_r);
+		bindHandler(Opcode.I64_ATOMIC_RMW_OR);
+		genAtomicBinop(asm.orq_m_r, asm.xchgq_m_r);
+		bindHandler(Opcode.I32_ATOMIC_RMW8_OR_U);
+		bindHandler(Opcode.I64_ATOMIC_RMW8_OR_U);
+		genAtomicBinop(asm.orb_m_r, asm.xchgb_m_r);
+		bindHandler(Opcode.I32_ATOMIC_RMW16_OR_U);
+		bindHandler(Opcode.I64_ATOMIC_RMW16_OR_U);
+		genAtomicBinop(asm.orw_m_r, asm.xchgw_m_r);
+		// Atomic RMW Xor
+		bindHandler(Opcode.I32_ATOMIC_RMW_XOR);
+		bindHandler(Opcode.I64_ATOMIC_RMW32_XOR_U);
+		genAtomicBinop(asm.xord_m_r, asm.xchgd_m_r);
+		bindHandler(Opcode.I64_ATOMIC_RMW_XOR);
+		genAtomicBinop(asm.xorq_m_r, asm.xchgq_m_r);
+		bindHandler(Opcode.I32_ATOMIC_RMW8_XOR_U);
+		bindHandler(Opcode.I64_ATOMIC_RMW8_XOR_U);
+		genAtomicBinop(asm.xorb_m_r, asm.xchgb_m_r);
+		bindHandler(Opcode.I32_ATOMIC_RMW16_XOR_U);
+		bindHandler(Opcode.I64_ATOMIC_RMW16_XOR_U);
+		genAtomicBinop(asm.xorw_m_r, asm.xchgw_m_r);
+		// Atomic RMW xchg
+		bindHandler(Opcode.I32_ATOMIC_RMW_XCHG);
+		bindHandler(Opcode.I64_ATOMIC_RMW32_XCHG_U);
+		genAtomicExchange(asm.xchgd_m_r);
+		bindHandler(Opcode.I64_ATOMIC_RMW_XCHG);
+		genAtomicExchange(asm.xchgq_m_r);
+		bindHandler(Opcode.I32_ATOMIC_RMW8_XCHG_U);
+		bindHandler(Opcode.I64_ATOMIC_RMW8_XCHG_U);
+		genAtomicExchange(asm.xchgb_m_r);
+		bindHandler(Opcode.I32_ATOMIC_RMW16_XCHG_U);
+		bindHandler(Opcode.I64_ATOMIC_RMW16_XCHG_U);
+		genAtomicExchange(asm.xchgw_m_r);
+		// Atomic RMW cmpxchg
+		bindHandler(Opcode.I32_ATOMIC_RMW_CMPXCHG);
+		bindHandler(Opcode.I64_ATOMIC_RMW32_CMPXCHG_U);
+		genAtomicCompareAndExchange(asm.d.cmpxchg_m_r);
+		bindHandler(Opcode.I64_ATOMIC_RMW_CMPXCHG);
+		genAtomicCompareAndExchange(asm.q.cmpxchg_m_r);
+		bindHandler(Opcode.I32_ATOMIC_RMW8_CMPXCHG_U);
+		bindHandler(Opcode.I64_ATOMIC_RMW8_CMPXCHG_U);
+		genAtomicCompareAndExchange(asm.cmpxchgb_m_r);
+		bindHandler(Opcode.I32_ATOMIC_RMW16_CMPXCHG_U);
+		bindHandler(Opcode.I64_ATOMIC_RMW16_CMPXCHG_U);
+		genAtomicCompareAndExchange(asm.cmpxchgw_m_r);
+    }
+    def genAtomicOp<T>(op: (X86_64Addr, X86_64Gpr) -> T,
```

Comment:
This method is getting a little big. In particular, it takes a `bool` that switches off a lot of the functionality. I think it makes sense to split it into two methods at least, and factor out some of the commonality to shared functions.

Reply:
Sure thing! I didn’t like it either tbh. But since it was done in a single function I kept like that to get feedback

Reply:
Ok, I am going to merge this and we can fix it in a followup.

Reply:
Hey! Thanks! I am going to make another PR about JIT and Virgil interpreters. I will write separate functions when making a PR about wait and notify o7